Add s3tokens validation.

chmouel opened the pull request:
There is still some works to be done in the swift middleware but this
should works.

Documentation and tests are in the planning.

chmouel committed:
Add s3tokens validation.

There is still some works to be done in the swift middleware but this
should works.

Documentation and tests are in the planning.

termie commented on keystone/contrib/s3/core.py:
```diff
@@ -0,0 +1,98 @@
+from keystone import token
+from keystone import service
+from keystone.common import wsgi
+from keystone.contrib.ec2 import Manager as EC2Manager
```
should import the module not the class, so:

from keystone.contrib import ec2

and later use ec2.Manager()

termie commented on keystone/contrib/s3/core.py:
```diff
@@ -0,0 +1,98 @@
+        self.identity_api = identity.Manager()
+        self.token_api = token.Manager()
+        self.policy_api = policy.Manager()
+        self.ec2_api = EC2Manager()
```
ec2.Manager()

termie commented on keystone/contrib/s3/core.py:
```diff
@@ -0,0 +1,98 @@
+        creds_ref = self.ec2_api.get_credential(context,
+                                                credentials['access'])
+
+        msg = base64.urlsafe_b64decode(str(credentials['token']))
```
could you make this little section a separate "sign_s3" function?

termie commented on keystone/contrib/s3/core.py:
```diff
@@ -0,0 +1,98 @@
+                context=context,
+                user_id=user_ref['id'],
+                tenant_id=tenant_ref['id'],
+                    metadata=metadata_ref)
```
extra indent

termie commented on keystone/contrib/s3/core.py:
```diff
@@ -0,0 +1,98 @@
+        self.ec2_api = EC2Manager()
+        super(S3Controller, self).__init__()
+
+    def authenticate_s3(self, context, credentials=None):
```
could this whole function wrap authenticate_ec2 and just change the signing? As in, make the signing part an instance method and have the s3 class override the signer

chmouel committed:
Simplify code.

Per Termie's suggestion, share as much as possible with ec2 module and
instantiate s3 module from it.

anotherjesse commented:
I've never liked the "ec2 token" name because the tokens are both AWS-style tokens....

chmouel commented:
What about having a /awstokens for tokens CRUD and /{ec2,s3}tokens for
specific services validation? everything will be under
keystone.contrib.awstokens  directory.

chmouel committed:
Make it as a subclass.

as advised by termie make it as a subclass instead of patching the
method.

termie commented on keystone/contrib/ec2/core.py:
```diff
@@ -83,7 +85,24 @@ def __init__(self):
+            raise Exception("Not Authorized")
+
+    def authenticate(self, context, credentials=None,
                          ec2Credentials=None):
```
could you re-indent this to match with self, pls

termie commented on keystone/contrib/ec2/core.py:
```diff
@@ -162,8 +171,9 @@ def authenticate_ec2(self, context, credentials=None,
-                self, token_ref, roles_ref, catalog_ref)
+        token_controller = service.TokenController()
+        return token_controller._format_authenticate(
+            token_ref, roles_ref, catalog_ref)
```
when moving to a new line for a long line, typically one double-indents the second line (as it was before this change)

chmouel replied:
> when moving to a new line for a long line, typically one double-indents
> the second line (as it was before this change)

emacs python-mode doesn't seem to like it much , oh well I'll use vim for
that one..

termie replied:
weird, pretty sure that one is in pep8's guide too

chmouel replied:
> weird, pretty sure that one is in pep8's guide too
> 
> FWIW: bin/pep8 command (0.6.1) doesn't seem to bug about it, but well
> anyway I have updated the commit ranges of this pull req with the
> indentations as they were before.

termie commented on keystone/contrib/ec2/core.py:
```diff
@@ -113,27 +132,17 @@ def authenticate_ec2(self, context, credentials=None,
-        tenant_ref = self.identity_api.get_tenant(creds_ref['tenant_id'])
-        user_ref = self.identity_api.get_user(creds_ref['user_id'])
+        tenant_ref = self.identity_api.get_tenant(
+            context=context,
```
could you double-indent or align these with the end of get_tenant(

chmouel committed:
indents.
